File: src/Sort/TopKFrequentElements/TopKFrequentElements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // save each element's appearance in a table
        Map<Integer, Integer> bucket1 = new HashMap<>();
        for (int num : nums) {
            bucket1.put(num, bucket1.getOrDefault(num, 0) + 1);
        }

        // Store each element's frequence as the key
        Map<Integer, List<Integer>> bucket2 = new HashMap<>();
        for (Integer num : bucket1.keySet()) {
            Integer elementFreq = bucket1.get(num);
            if (!bucket2.containsKey(elementFreq)) {
                bucket2.put(elementFreq, new ArrayList<>());
            }
            bucket2.get(elementFreq).add(num);
        }
        // Get Top K elements
        int[] res = new int[k];
        for (int n = nums.length; n > 0; n--) {
            if (bucket2.containsKey(n)) {
                List<Integer> list = bucket2.get(n);
                for (Integer integer : list) {
                    res[--k] = integer;
                    if (k == 0)
                        return res;
                }
            }
        }
        return res;
    }
}

File: src/Sort/TopKFrequentElements/TopKFrequentElements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Store each element's frequency
        freq_by_num = dict()
        max_freq = 0
        for num in nums:
            if num not in freq_by_num:
                freq_by_num[num] = 0
            freq_by_num[num] += 1
            max_freq = max(max_freq, freq_by_num[num])
                
        num_by_freq = dict()
        for num in freq_by_num:
            freq = freq_by_num[num]
            if freq not in num_by_freq:
                num_by_freq[freq] = []
            num_by_freq[freq].append(num)
        
        # Find the top k frequent element
        result = []
        
        for cur_freq in range(max_freq, -1, -1):
            if cur_freq in num_by_freq:
                cur_nums = num_by_freq[cur_freq]
                for num in cur_nums:
                    result.append(num)
                    k -= 1
                    if k == 0: 
                        return result
